includes/class-wc-google-gtag-js.php

```diff
@@ -37,498 +36,185 @@ public static function get_instance( $options = array() ) {
-			$categories = get_the_terms( $product->get_parent_id(), 'product_cat' );
-		}
-
-		if ( false !== $categories && ! is_wp_error( $categories ) ) {
-			foreach ( $categories as $category ) {
-				$category_names[] = $category->name;
-			}
-		}
-
-		return join( '/', $category_names );
-	}
-
-	/**
-	 * Return list name for event
-	 *
-	 * @return string
-	 */
-	public static function get_list_name(): string {
-		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
-		return isset( $_GET['s'] ) ? __( 'Search Results', 'woocommerce-google-analytics-integration' ) : __( 'Product List', 'woocommerce-google-analytics-integration' );
 	}
 
 	/**
-	 * Enqueues JavaScript to build the view_item_list event
+	 * Add an event to the script data
 	 *
-	 * @param WC_Product $product
-	 */
-	public static function listing_impression( $product ) {
-		$event_code = self::get_event_code(
-			'view_item_list',
-			array(
-				'items' => array(
-					array(
-						'id'       => self::get_product_identifier( $product ),
-						'name'     => $product->get_title(),
-						'category' => self::product_get_category_line( $product ),
-						'list'     => self::get_list_name(),
-					),
-				),
-			)
-		);
-
-		wc_enqueue_js( $event_code );
-	}
-
-	/**
-	 * Enqueues JavaScript for select_content and add_to_cart events for the product archive
+	 * @param string       $type The type of event this data is related to.
+	 * @param string|array $data The event data to add.
+	 * @param string       $key  If not null then the $data will be added as a new array item with this key.
+	 * @param bool         $single If true then no other data will be added for this type.
 	 *
-	 * @param WC_Product $product
-	 */
-	public static function listing_click( $product ) {
-		$item = array(
-			'id'       => self::get_product_identifier( $product ),
-			'name'     => $product->get_title(),
-			'category' => self::product_get_category_line( $product ),
-			'quantity' => 1,
-		);
-
-		$select_content_event_code = self::get_event_code(
-			'select_content',
-			array(
-				'items' => array( $item ),
-			)
-		);
-
-		$add_to_cart_event_code = self::get_event_code(
-			'add_to_cart',
-			array(
-				'items' => array( $item ),
-			)
-		);
-
-		wc_enqueue_js(
-			"
-			$( '.product.post-" . esc_js( $product->get_id() ) . ' a , .product.post-' . esc_js( $product->get_id() ) . " button' ).on('click', function() {
-				if ( false === $(this).hasClass( 'product_type_variable' ) && false === $(this).hasClass( 'product_type_grouped' ) ) {
-					$add_to_cart_event_code
-				} else {
-					$select_content_event_code
-				}
-			});"
-		);
-	}
-
-	/**
-	 * Output Javascript to track add_to_cart event on single product page
+	 * @return void
 	 */
-	public static function add_to_cart() {
-		if ( 'yes' !== self::get( 'ga_event_tracking_enabled' ) || ! is_single() ) {
-			return;
+	public function set_script_data( string $type, $data, ?string $key = null, bool $single = false ): void {
```

Comment:
:nail_care: 
I find this API a bit confusing, as if I understood correctly it's like:
```
set( where1, what, [where2], [where_modifier])
```
WDYT of something like:

```
set( what, where1, [where2], [where_modifier])
```
so:
```suggestion
	public function set_script_data( $data, string $type, ?string $key = null, bool $single = false ): void {
```

Reply:
Also, the majority of useage does: `set_script_data( *, *, null, true)`, and it would behave the same like `set_script_data( *, *, "this-argument-is-pointless-now", true)`. 

So maybe we could make the "overwrite the root property" the default. Change the API, to something like:
`set_script_data( *, *, ?string $key_or_push = null ): void {`

```php
	/**
	 * Add an event to the script data
	 *
	 * @param string           $type The type of event this data is related to.
	 * @param string|array     $data The event data to add.
	 * @param string|bool|null $key  Set to add the data as a new array item, `true` to push with a numeric key.
	 *
	 * @return void
	 */
	public function set_script_data( string $type, $data, ?string $key = null ): void {
	// or public function set_script_data( $data, string $type, ?string $key = null ): void {
		if ( is_null( $key ) ) {
			$this->script_data[ $type ] = $data;
		} else {
			if ( ! isset( $this->script_data[ $type ] ) ) {
				$this->script_data[ $type ] = array();
			}
			if( $key === true ) {
				$this->script_data[ $type ][] = $data;
			} else {
				$this->script_data[ $type ][ $key ] = $data;
			}
		}
	}
```

Then the calls would look like:
- overwrite the root: 
   ```php
   $this->set_script_data( *, * );
   $this->set_script_data( 'cart', $this->get_formatted_cart() );
   ```
- specify key: 
   ```php
   $this->set_script_data( *, *, 'key' );
   $this->set_script_data( 'events', $gtag_event, $gtag_event );
   ```
- push to array: 
   ```php
   $this->set_script_data( *, *, true );
   $this->set_script_data( 'products', $this->get_formatted_product( $product ), true );
   ```

Reply:
Thanks for highlighting that @tomalec. It evolved over time and you're right it was structured in a fairly confusing way. I've instead simplified how it works and added a separate function to manage adding more data in c4931e3 (_Related to the [comment above](https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/328#discussion_r1484587515)_).